### ISSUE TITLE
[FlatList] FlatList crashes on device

### ISSUE THREAD

**Issue by ishrak-imam** [facebook/react-native#14086]
### Description

I request server for data 10 items at a time and render those in a FlatList component. Everything is working fine in the emulator but when I build the app in device and scroll up and down very fast the application crashes. I used pure component for each row item of the list and implemented a getItemLayout method but the result is still same.

### Example code

import React, { Component } from "react";
import { Text, FlatList, View } from "react-native";
import HistoryRow from './historyRow'
import Loader from '../components/listLoader'
import styles from '../styles/style'

class HistoryList extends Component {
	constructor(props) {
		super(props)
	}

	renderSeparator = () => {
		return (
			<View style={{ height: 1, width: "100%", backgroundColor: "#CED0CE", }} />
		);
	};

	renderFooter = () => {
		if (!this.props.loading) return null;
		return (
			<View style={{ paddingVertical: 20, borderTopWidth: 1, borderColor: "#CED0CE" }} >
				<Loader></Loader>
			</View>
		);
	};

	render() {

		const { data, refreshCallback, loadMoreCallback, refreshing, loading, historyEnd } = this.props

		if(data.length === 0 && !loading) {
			return (
				<View style={[styles.container, {marginTop: 50}]}>
					<View style={[styles.colCenterCenter]}>
						<Text>No data found</Text>
					</View>
				</View>
			)
		}

		return (
			<FlatList
				data={data}
				renderItem={({ item }) => (
					<HistoryRow data={item} ></HistoryRow>
				)}

				getItemLayout={(data, index) => {
					return { length: 71, offset: 71 * index, index: index }
				}}

				removeClippedSubviews={false}

				keyExtractor={item => item.transactionId}
				ItemSeparatorComponent={this.renderSeparator}
				ListFooterComponent={this.renderFooter}
				onRefresh={() => { refreshCallback() }}
				refreshing={refreshing}
				onEndReached={() => { loadMoreCallback() }}
				onEndReachedThreshold={.5}
			/>
		);
	}
}
export default HistoryList;

this is the FlatList component above

//import libraries
import React, { Component } from 'react'
import { View, Text, TouchableHighlight } from 'react-native'
import styles from '../styles/style'

// create a component
export default class HistoryRow extends React.PureComponent {
	
	// pure component implements shouldComponentUpdate by default
	constructor(props) {
		super(props)
		this.state = {
			color: 'black'
		}
	}

	// shouldComponentUpdate(nextProps) {
	// 	return this.props !== nextProps
	// }

	colorToggle = (color) => {
		this.setState({color})
	}

	render() {

		const { data } = this.props

		return (
			<TouchableHighlight
				style={{ paddingVertical: 15, paddingHorizontal: 20, height: 70 }}
				underlayColor="#f47727"
				onPress={() => { }}
				onShowUnderlay={() => { this.colorToggle('white')}}
				onHideUnderlay={() => { this.colorToggle('black')}}
			>
				<View style={[styles.flex1, styles.colSpaceAroundStart]}>
					<View style={[styles.flex1]}>
						<View style={[styles.flex1, styles.rowCenterCenter]}>
							<Text style={{color: this.state.color}}>Date: {data.transactionDate}</Text>
							<Text style={{ marginLeft: 30, color: this.state.color }}>Coin: {data.coinAmount}</Text>
						</View>
					</View>
					<View style={[styles.flex2]}>
						<View style={[styles.flex1, styles.colCenterCenter]}>
							<Text style={{color: this.state.color}}>Description: {data.transactionRemarks}</Text>
						</View>
					</View>
				</View>
			</TouchableHighlight>
		)
	}
}

and here is the row item component.

in the container component inside the componentDid mount i do server request to get the data and I use simple setState when data is received. I am not using redux and I don't think the issue can be related to usage of redux. and after first server call i request for new data inside the onEndReached callback and as the new data comes I merge then with the previous list and call setState again. this is roughly the cycle and it works in emulator even in android device in debug mode. But when I generate a release apk and install it in device not always after some scrolling up and down the application gets crashed.

I am using 
classpath 'com.android.tools.build:gradle:1.2.3'
distributionUrl=https\://services.gradle.org/distributions/gradle-2.2-all.zip
buildToolsVersion "23.0.2"

**Comment by windhost** [comment]
I got the same problem. The error is:
`java.lang.IllegalStateException: The specified child already has a parent. You must call removeView() on the child's parent first.
                                                            at android.view.ViewGroup.addViewInner(ViewGroup.java:3562)
                                                            at android.view.ViewGroup.addView(ViewGroup.java:3415)
                                                            at android.view.ViewGroup.addView(ViewGroup.java:3360)
                                                            at com.facebook.react.views.view.ReactViewManager.addView(ReactViewManager.java:209)
                                                            at com.facebook.react.views.view.ReactViewManager.addView(ReactViewManager.java:41)
                                                            at com.facebook.react.uimanager.NativeViewHierarchyManager.manageChildren(NativeViewHierarchyManager.java:394)
                                                            at com.facebook.react.uimanager.UIViewOperationQueue$ManageChildrenOperation.execute(UIViewOperationQueue.java:179)
                                                            at com.facebook.react.uimanager.UIViewOperationQueue$2.run(UIViewOperationQueue.java:787)
                                                            at com.facebook.react.uimanager.UIViewOperationQueue.flushPendingBatches(UIViewOperationQueue.java:843)
                                                            at com.facebook.react.uimanager.UIViewOperationQueue.access$1600(UIViewOperationQueue.java:48)
                                                            at com.facebook.react.uimanager.UIViewOperationQueue$DispatchUIFrameCallback.doFrameGuarded(UIViewOperationQueue.java:889)
                                                            at com.facebook.react.uimanager.GuardedFrameCallback.doFrame(GuardedFrameCallback.java:31)
                                                            at com.facebook.react.modules.core.ReactChoreographer$ReactChoreographerDispatcher.doFrame(ReactChoreographer.java:129)
                                                            at com.facebook.react.modules.core.ChoreographerCompat$FrameCallback$1.doFrame(ChoreographerCompat.java:107)
                                                            at android.view.Choreographer$CallbackRecord.run(Choreographer.java:759)
                                                            at android.view.Choreographer.doCallbacks(Choreographer.java:574)
                                                            at android.view.Choreographer.doFrame(Choreographer.java:543)
                                                            at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:747)
                                                            at android.os.Handler.handleCallback(Handler.java:733)
                                                            at android.os.Handler.dispatchMessage(Handler.java:95)
                                                            at android.os.Looper.loop(Looper.java:136)
                                                            at android.app.ActivityThread.main(ActivityThread.java:5001)
                                                            at java.lang.reflect.Method.invokeNative(Native Method)
                                                            at java.lang.reflect.Method.invoke(Method.java:515)
                                                            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:785)
                                                            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:601)
                                                            at dalvik.system.NativeStart.main(Native Method)`

**Comment by windhost** [comment]
I fixed this by remove stickySectionHeadersEnabled option in android.

**Comment by annelorraineuy** [comment]
My FlatList crashes as well, more so after upgrading to RN 0.44.0

**Comment by peacechen** [comment]
This is a serious bug that should be given high priority.  We had observed crashes with large static lists in RN 0.34's ScrollView and ListView.  I optimistically upgraded to RN 0.45.1 to use FlatList which is claimed to be more stable.  In our limited testing I haven't seen a crash with large static lists of items in FlatList, but it has crashed when dynamically changing the list.

Use case is filtering the items in a FlatList.  On initial render all items are shown, then results may be pared down through a search bar.

It did seem to crash when scrolling through the list, but I'm not sure if it was in the process of filtering the results.

**Comment by peacechen** [comment]
It's crashing when rotating the screen.  Every 1-3 rotations will consistently crash the app.

**Comment by peacechen** [comment]
@GantMan @superyarik @reneweb @sahrens 
Anyone know who maintains FlatList?  The crash on rotation is a show stopper.

**Comment by peacechen** [comment]
@ericvicenti  Would you be able to help us out?

**Comment by sahrens** [comment]
It's pretty hard to diagnose or even triage the issue without a stack trace of the crash...can you please add one? It's not clear if it's even a native crash or JS exception from your description. Also a repro example would be great, ideally with expo sketch.

**Comment by peacechen** [comment]
Unfortunately I can't capture a stack trace because it's a silent crash.  The app quits entirely and there's nothing in the console log.  It's much harder to get it to crash in the debugger, yet is very reproducible in a production build.  Try rotating the device a few times with a populated FlatList.

**Comment by sahrens** [comment]
There should be someway to get the output. This may help:

https://stackoverflow.com/questions/35364373/android-app-crashes-without-any-logcat-or-any-exception

At the very least please include the full logcat output in a gist or something.

**Comment by peacechen** [comment]
Thanks that's helpful.  Is there an equivalent for iOS / Xcode?

**Comment by sahrens** [comment]
Another google search reveals this: https://stackoverflow.com/questions/36788837/ios-app-crashes-but-there-is-no-crash-log-available

How are you running and installing your production build? Are you using expo? If not, what native modules are you using? The more details you provide, the more likely we can help.

**Comment by sahrens** [comment]
Note that this is probably not a FlatList issue - JS code should not be able to crash the device without a JS stack trace. It's probably a bug with the bridge, layout, some native module or native glue code, and your list is probably tickling the bug somehow.

**Comment by peacechen** [comment]
I've still not been able to capture any information in the debugger.  One crash report pointed to Layout Animation which is used in the FlatList rows.  Not definitive whether that's the culprit. Will continue investigating.

**Comment by MalguyMQ** [comment]
Same issue but on old device particularly. I noticed that on android over 4 years. I'm using redux too. Without Redux, crash is less frequent..

**Comment by eli-isbell-h360** [comment]
In case this benefits anyone, won't apply to all... check for seemingly non-related culprits. As @sahrens pointed out, problem likely isn't a FlatList issue. In my case I found a custom svg icon in my renderItem component was causing the crash on device when it hadn't elsewhere—replaced it with a react-native-vector-icon and everything functions as it should.

**Comment by peacechen** [comment]
I'm using Redux as well.  I'm not convinced this isn't a Flatlist bug.  Definitely a corner case that's very hard to reproduce in the debugger.  The reasons I believe it's Flatlist:

- The exception happens in native code. No JS stacktrace.
- No crashes when the list is small (~10) even with complex rows.
- Crashes as the number of rows and complexity increases.  Empirically this has been around 50+ rows.

My hunch is that it's related to FlatList dynamically culling the off-screen rows from memory and adding new ones as they come into view.  Many of the NewRelic crash reports end with `libobjc.A.dylib | objc_object::release`

**Comment by neomib** [comment]
I'm using mobx and my list crashes when the app navigates to a new screen after the user clicked on one of the rows. When there is a small amount of rows everything works perfect!
Any update on this?

**Comment by phthhieu** [comment]
Does it work after removing `ItemSeparatorComponent`? Im facing this recently

**Comment by peacechen** [comment]
Our app doesn't use `ItemSeparatorComponent` in FlatList.

I have reduced the crash frequency by a small amount by minimizing state updates.  Standard React optimization strategies using shouldComponentUpdate, componentWillReceiveProps, etc.  Hopefully that gives a clue to the underlying cause.

**Comment by MalguyMQ** [comment]
Best way is to never update state in the flatlist. Only update props with
extraData on props who shouldUpdate and your flatlist will never crash.
This is a pure component, that’s why :)

**Comment by pavanmehta91** [comment]
@peacechen  Do I need pure component even if my flat list does not change. By state updates are you talking about overall state updates or state updates for the dataSource  used in flatlist?

**Comment by peacechen** [comment]
Thanks @MalguyMQ , I'll look into how state updates are happening.

@pavanmehta91  yes I believe state updates are being passed to dataSource.

**Comment by peacechen** [comment]
I've updated it so that the data prop passed to `FlatList` never changes.  The `renderItem` method applies the filter.
This unfortunately doesn't work in another place where `numColumns = 2`.  `renderItem` applies the filter and returns undefined for items that are filtered out, but FlatList places an undesired blank space there.

@MalguyMQ  How does one implement infinite scroll if FlatLists's `data` props is never mutated?  This guy freely mutates state passed to data prop:
https://www.youtube.com/watch?v=rY0braBBlgw

**Comment by MalguyMQ** [comment]
@peacechen, the best way to filter your data is to call a function who filter data in your props and dispatch new data filtered. If you are using redux. If you don’t use redux of external state you need to use your state inside your component. You can did it in your state if you want. But don’t forget in the video it’s just an exemple, not ready for production. 

Last advice, when your props it’s updated, if your extraData is like: extraData={this.props} and your dataSource={this.props.myData} your render items function will be called for each new items. 

Can you share your code if your not able to achieve this. It’s gonna be more easy to help you ;)

**Comment by phantom1299** [comment]
Hi, I'm having the same issue. App crashes after pressing a Flatlist element with onPress prop.
I only have that issue when remote JS debbuging is off. When its on, everything works fine.
I have TouchableOpacity as Parent in all my rows in Flatlist. It crashes only when TouchableOpacity has onPress prop with a function other then () => {}. When I dont provide an onPress prop or set it to () => {}, it doesnt crash at all. 

Tryed with TouchableWithoutFeedback and others too, same result.

**Comment by nikolal** [comment]
FlatList is crashing for me only if I set horizonta={true}, it never crashes without it.

**Comment by jamesreilly5** [comment]
I've just run into the same issue when I upgraded from ListView to FlatList. I have an infinite scroll which loads 10 items at a time. First page is fine and then after the second page loads it works for a second and then silently crashes.

All I can see in the logs is an "out of memory error", no other logs indicating the root cause. I only got a little time yesterday to look into it so will spend some time time tonight trying to diagnose. If I find my problem I'll update (or if it's unrelated to FlatList I'll delete this comment).

**Comment by jordangrant** [comment]
Might not help others but, my FlatList caused silent crashes when I was calling:
```
componentWillUnmount() {
    RNFetchBlob.fs.unlink(this.state.path);
  }
```
for every FlatList element. When swipe to go back was initiated, it would cause the crash on the large list ~30 items.

**Comment by hugoh1995** [comment]
Both my scrollview with netsted flatlists and regular scrollview crashes on Android devices only.

**Comment by neomib** [comment]
For me, the problem was passing large objects when navigating. When I changed it to strings and saved everything in a service, everything worked fine.

**Comment by pavanmehta91** [comment]
My problem gotten resolved I did 2 things
1) Move my FlatList outside from the ScrollView (Note: It only crashed in android and not in IOS)
2) I resized my images before storing them in s3.

**Comment by jamesreilly5** [comment]
Ok I found that my issue was not related to FlatList at all, it was two things and both were related to my SVGs. I found this out by profiling the memory usage using `Instruments` while using the device (iPad) so if anyone is having trouble I'd advise you to see if it's one of your components causing a leak first.

My issue was first of all a memory leak issue with [react-native-svg](https://github.com/react-native-community/react-native-svg/issues/278) which has since been fixed. My other issue was that making SVGs that are dimensions similar to the device size eat **huge** amounts of memory (I would have thought as an SVG setting the width/height wouldn't matter as it's an SVG but oh well). I fixed my memory issues by making the SVGs really small and using transform's scale property instead. I hope this helps anyone that had a similar issue. Obviously not saying there aren't edge case crashes which FlatList might have
